### ISSUE TITLE
TypeError: this function takes at least 3 arguments (0 given)

### ISSUE THREAD

**Issue by Fahmie23** [guidance-ai/guidance#659]
This is my code

<img width="1043" alt="image" src="https://github.com/guidance-ai/guidance/assets/130896959/eedc5673-4b1c-4ccd-900b-134ce0ef9b7b">

This is the error that occur during trying to run the code
<img width="1180" alt="image" src="https://github.com/guidance-ai/guidance/assets/130896959/bbf47edc-1df8-4d3a-9c90-401ef02eb5fc">

**Comment by jtha** [comment]
I'm seeing the same error after reinstalling my llama-cpp-python library, no issues before.

**Comment by jtha** [comment]
@Harsha-Nori 

guidance                                       0.1.10
llama_cpp_python                         0.2.52

**Comment by jtha** [comment]
@Fahmie23 @Harsha-Nori  I finally got this to work. I had to make a few changes to `_llama_cpp.py`

class _LlamaBatchContext:
    def __init__(self, n_batch, n_ctx):
        self._llama_batch_free = llama_cpp.llama_batch_free
        `self.batch = llama_cpp.llama_batch_init(n_batch, 0, n_ctx)`
        if self.batch is None:
            raise Exception("call to llama_cpp.llama_batch_init returned NULL.")

and further down below, changed all the `llama` parts from e.g.
`self._n_vocab = tokenizer.llama.n_vocab()`
to 
`self._n_vocab = tokenizer._model.n_vocab()`

**Comment by danielvarab** [comment]
I'm facing with this as well. The response @jtha wrote only shifted prolonged some errors. Is this a llama_cpp problem? I've tried out several versions of llama_cpp and not avoiding the error.

**Comment by danielvarab** [comment]
update: managed to fix it by reinstalling `guidance` (latest) and `llama-cpp-python==0.2.26`.

This suggests that there are compatibility issues with newer versions of `llama-cpp-python`. Also mentioned in #608

**Comment by Fahmie23** [comment]
@danielvarab Yes correct, I also manage to fix it with the same method as you.

**Comment by jtha** [comment]
good call on identifying the compatible llama-cpp version @danielvarab, you were correct I eventually ran into issues if loading the chat model (seems tokenizer not set correctly). ended up reinstalling to 0.2.26

**Comment by antonsapt4** [comment]
Hi guys, had the same problem `TypeError: this function takes at least 3 arguments (0 given)`  using chat model

tried 
@jtha method change `self._n_vocab = tokenizer._model.n_vocab()`

@danielvarab reinstall (even use force reinstall flag) guidance and make sure the llama-cpp-python is on 0.2.26

But the issue still remain wondering if how to resolved it or there is something wrong with my code

Guidance Version: 0.1.10
llama-cpp-python Version: 0.2.26
python version: 3.11.8

```python
model_path = "/Users/aldi/.cache/lm-studio/models/TheBloke/OpenHermes-2.5-Mistral-7B-GGUF/"
openhermes = models.LlamaCppChat(model=model_path, n_gpu_layers=-1, n_ctx=4096)

lm = openhermes
system_message = 'You are a helpful assistant'
user_message = 'I need a coffee'

with system():
    lm + system_message

with user():
    lm += user_message

with assistant():
    lm += gen(max_tokens=10)
```
